### PR TITLE
Lint for invisible Unicode characters other than ZWSP

### PR DESCRIPTION
**Pull request by bugadani**
This PR extends the existing `zero_width_space` lint to look for other invisible characters as well (in this case, `\\u{ad}` soft hyphen.

I feel like this lint is the logical place to add the check, but I also realize the lint name is not particularly flexible, but I also understand that it shouldn't be renamed for compatibility reasons.

Open questions:
 - What other characters should trigger the lint?
 - What should be done with the lint name?
 - How to indicate the change in functionality?

Motivation behind this PR: https://github.com/rust-lang/rust/issues/77417 - I managed to shoot myself in the foot by an invisible character pasted into my test case.

changelog: rename [`zero_width_space`] to [`invisible_characters`] and add SHY and WJ to the list.

### CODE REVIEW

File: clippy_lints/src/unicode.rs [hunk 1]
```diff
@@ -91,14 +91,14 @@ fn escape<T: Iterator<Item = char>>(s: T) -> String {
...
 fn check_str(cx: &LateContext<'_>, span: Span, id: HirId) {
     let string = snippet(cx, span, "");
-    if string.contains('\u{200B}') {
+    if let Some(invisible) = string.chars().find(|c| ['\u{200B}', '\u{ad}'].contains(&c)) {
         span_lint_and_sugg(
             cx,
             ZERO_WIDTH_SPACE,
             span,
-            "zero-width space detected",
+            &format!("invisible character detected: {:?}", invisible),
```
Comment (ebroto) [inline clippy_lints/src/unicode.rs:99]: As we are going to replace all occurrences, I think we should either specify all the characters that are going to be replaced, or just say `invisible character detected` without specifying which one (they will be visible in the suggestion anyway)

**Review by ebroto — CHANGES_REQUESTED** [review]
About the open questions:

>What other characters should trigger the lint?

I'm *really* not an expert in this field. Taking a look at  [this Wikipedia entry](https://en.wikipedia.org/wiki/Whitespace_character#Unicode), there are some other invisible characters (second table) which could be candidates. 
My main concern would be including characters that could be legitimately used in languages I'm not familiar with, so I would prefer to be conservative here. Maybe the `WORD JOINER` one could be a candidate? If we are not sure I would prefer not to include others.

>What should be done with the lint name?

There's a mechanism for renaming lints. You can find an example in #3554. It's a bit old PR, but self-contained, except for the test about renaming lints, you can omit that.

>How to indicate the change in functionality?

A new name for the lint should suffice. What about `invisible_characters` that you used when linting?

**Comment by bugadani** [comment]
> I'm _really_ not an expert in this field. Taking a look at [this Wikipedia entry](https://en.wikipedia.org/wiki/Whitespace_character#Unicode), there are some other invisible characters (second table) which could be candidates.

Oh, I intended this question to a broader audience, not placing the burden of figuring out on you. Just in case somebody would have a stronger opinion here.

> My main concern would be including characters that could be legitimately used in languages I'm not familiar with, so I would prefer to be conservative here. Maybe the `WORD JOINER` one could be a candidate? If we are not sure I would prefer not to include others.

I've added WJ, that seems obvious enough. I don't think language support is an issue here - if a character is invisible, I think people should prefer writing out the escape code instead in any language. But I may have wrong assumptions and preferences here.

> There's a mechanism for renaming lints. You can find an example in #3554. It's a bit old PR, but self-contained, except for the test about renaming lints, you can omit that.

Thanks, I've renamed the lint to `invisible_characters`.

File: clippy_lints/src/unicode.rs (hunk 1 above)
Comment (bugadani) [reply clippy_lints/src/unicode.rs]: Fair enough, thanks!

**Review by ebroto — APPROVED** [review]
LGTM, let's wait for CI and we can merge it.

>Oh, I intended this question to a broader audience, not placing the burden of figuring out on you

Of course! Sorry if that sounded harsh, it was not the intention.

>I don't think language support is an issue here - if a character is invisible, I think people should prefer writing out the escape code instead in any language

That makes sense, but I would prefer to be cautious. I think adding those two is a good improvement in any case.

**Comment by ebroto** [comment]
@bors r+

Thanks!

**Comment by bugadani** [comment]
> Of course! Sorry if that sounded harsh, it was not the intention.

Not at all, no worries. :) Thanks for the review!

**Pull request merged**